File: src/ai/assistant.service.ts
import { LLMService } from "../shared/llm.service";

export class AssistantService {
  private llm: LLMService;

  constructor() {
    this.llm = new LLMService();
  }

  async run(message: string, history: any[]) {
    const prompt = `
You are an AI project consultant.

Your job is to:
1. Continue the conversation with the user.
2. Extract project information from the conversation.

Return STRICT JSON ONLY.

Format:
{
  "reply": "assistant reply to continue conversation",
  "project": {
    "projectType": "",
    "platform": "",
    "features": [],
    "budgetMin": null,
    "budgetMax": null,
    "timeline": ""
  }
}

Conversation history:
${JSON.stringify(history)}

User message:
${message}
`;

    const raw = await this.llm.call([{ role: "user", content: prompt }]);

    // Clean markdown if AI returns ```json
    const cleaned = raw
      .replace(/```json/g, "")
      .replace(/```/g, "")
      .trim();

    try {
      const parsed = JSON.parse(cleaned);

      return {
        reply: parsed.reply || "",
        project: parsed.project || {}
      };

    } catch (error) {

      console.error("❌ JSON parse failed");
      console.error(cleaned);

      return {
        reply: raw,
        project: {}
      };
    }
  }
}

File: src/ai/assistant_service.py
# ============================================================
# PATH: backend / ai / assistant_service.py
# PURPOSE: Standalone assistant — conversation + extraction in one call
# ============================================================

import json
import re
from typing import Any, Dict, List


class AssistantService:
    def __init__(self, llm_service):
        self.llm = llm_service

    async def run(self, message: str, history: List[Dict[str, Any]]) -> Dict[str, Any]:
        prompt = f"""
You are an AI project consultant.

Your job is to:
1. Continue the conversation with the user.
2. Extract project information from the conversation.

Return STRICT JSON ONLY.

Format:
{{
  "reply": "assistant reply to continue conversation",
  "project": {{
    "projectType": "",
    "platform": "",
    "features": [],
    "budgetMin": null,
    "budgetMax": null,
    "timeline": ""
  }}
}}

Conversation history:
{json.dumps(history)}

User message:
{message}
"""

        raw = await self.llm.call([{"role": "user", "content": prompt}], task="conversation")
        cleaned = re.sub(r"```json|```", "", raw, flags=re.IGNORECASE).strip()

        try:
            parsed = json.loads(cleaned)
            return {
                "reply": parsed.get("reply", ""),
                "project": parsed.get("project", {}),
            }
        except Exception:
            print("❌ JSON parse failed")
            print(cleaned)
            return {"reply": raw, "project": {}}